### ISSUE TITLE
v7 - assumption of global "webdriverio" is breaking the types

### ISSUE THREAD

**Issue by dylang** [webdriverio/webdriverio#6412]
**Environment (please complete the following information):**
 - **WebdriverIO version:** 7.0.2
 - **Mode:** Standlone, async
 - **Node.js version:** 14
 - **NPM version:** yarn 1
 - **Browser name and version:** n/a
 - **Platform name and version:** TypeScript 4.1.5
 - **Additional wdio packages used (if applicable):** DevTools 7.0.2

**Config of WebdriverIO**

Hi! We're using webdriverio standalone async with DevTools so we can run it from Jest with our other tests.

We use webdriver.io in many large projects where it would not make sense to make `webdriverio` a global.

**Describe the bug**

https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/commands/browser/url.ts#L38

`WebdriverIO` (the type) is not imported anywhere in this file, so the generated type does not import it. This means when this file is used in any project that does not have `webdriverio` as a global, it fails or returns `any` for `WebdriverIO.Browser`.

**To Reproduce**

```json5
// tsconfig.json
"types": [
  "node",
  "webdriverio/sync" <-- do not include this
],
```

```ts
import { remote } from 'webdriverio';

const tempTest = async () =>
  const webdriverioTemp = await remote(config);
  await webdriverioTemp.url('about:blank');

  return webdriverioTemp;
};
```

Hover over `.url`, incorrect type:

<img width="436" alt="Screen Shot 2021-02-11 at 4 22 39 PM" src="https://user-images.githubusercontent.com/51505/107700744-ef873d80-6c85-11eb-90ea-28e7b03f8a15.png">

**Expected behavior**

Correctly typed without expectation to make Webdriverio global.

**Log**

Enabled eslint rule `@typescript-eslint/no-unsafe-call` and it errors all over the place because of the `any` return types.

**Comment by christian-bromann** [comment]
> Correctly typed without expectation to make Webdriverio global.

We need to make it global to allow sync and async types depending on your settings. Why can't you set
```json
"types": [
  "node",
  "webdriverio/async" <-- do not include this
],
```

in your tsconfig?

**Comment by dylang** [comment]
Hi @christian-bromann, thanks for taking a look into this.

`webdriverio` is not a global variable in our projects. For example, `webdriverio` is not available in any frontend or backend code. It's also not available to our unit testing code, just the browser tests.  We always import it where we use it.

`node` isn't global is many projects either, such as any frontend code. 

If you look at the compiled types, TypeScript fixes this by adding:

```ts
/// <reference types="node" />
```

to files that reference node globals, such as https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/types.ts. 

It might be possible to add that to all files that reference `webdriverio` to work around this if you don't want to `import type { ... } from ...`:

```ts
/// <reference types="webdriverio/async" />
```

**Comment by dylang** [comment]
I'm giving what you suggested a try and it's not working for the same reason:

```ts
"types": [
  "node",
  "webdriverio/async"
],
```

```ts 
const el = await this.webdriverio.$(selector);
```

`$` returns `Promise<WebdriverIO.Element>;` but using the `types` above doesn't appear to help:

<img width="480" alt="Screen Shot 2021-02-14 at 4 49 39 PM" src="https://user-images.githubusercontent.com/51505/107890059-d3350c00-6ee4-11eb-8d96-96aff13b54a0.png">

```ts
export default function $(this: WebdriverIO.Browser, selector: Selector): Promise<WebdriverIO.Element>;
```

Even with `types` above, `WebdriverIO.Element` is not resolved.

My guess is `types` is only used to augment _my_ code, not _everything in node_modules_, including `webdriverio` package.

**Comment by christian-bromann** [comment]
@dylang can you provide a minimal reproducible example of your setup? This makes is easier to understand what you are experiencing.

Note that apart from other projects WebdriverIO needs to support a synchronous type set and an asynchronous one. The way we have implemented this is through injecting WebdriverIO as global namespace and always reference the namespace (e.g. `WebdriverIO.Browser`) when it can be either of them.

**Comment by dylang** [comment]
After struggling a bunch, I found some ways to get `types` working in my project!

Here's the problems and solution for others struggling with this as well:

* `types` should go the `compilerOptions` section of the `tsconfig.json`, not the top level where I had it at first.
* I'm using a monorepo and only one of the projects has `webdriverio` as a dependency, so the other packages in the monorepo were getting errors that tsc couldn't find `webdriverio/async`. This was solved by making sure only the package using webdriverio had `webdriverio/async` in `types`, and not the repo root's tsconfig.json.
* [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint) was using just the repo's tsconfig.json, instead of all of the tsconfigs. This change fixed that:
```ts
// Use all the tsconfig.json's
parserOptions: { project: ['./packages/**/tsconfig.json', './tsconfig.json'], tsconfigRootDir: process.cwd() },
```
* Close and re-open the project in WebStorm between `tsconfig.json` changes, otherwise the built-in `eslint` wasn't picking up the changes to the `tsconfig.json`.
* I also noticed `waitUntil` is hardcoded to return `Promise<any>`. @christian-bromann -  Maybe this should return `Promise<void>` because either it waits or it throws an error, we shouldn't care about what it returns. If we do, then maybe a generic for the return value? This was throwing me off because I thought the `any` was related to the original issue. https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/commands/browser/waitUntil.ts#L71

😅 

So in conclusion, this ticket can be closed as it is not a problem with webdriverio. Thanks for reading!

**Comment by christian-bromann** [comment]
> * `types` should go the `compilerOptions` section of the `tsconfig.json`, not the top level where I had it at first.

Yes, usually IDEs should give you an error if the tsconfig is wrongly typed

> * Maybe this should return `Promise<void>` because either it waits or it throws an error, we shouldn't care about what it returns.

Good point, creating an issue for it
